lib/DBus/Socket.hs

```diff
@@ -323,46 +324,74 @@ toSocketError addr io = catches io handlers where
 authenticator :: Authenticator t
 authenticator = Authenticator (\_ -> return False) (\_ _ -> return False)
 
+data UnixFdSupport = UnixFdsSupported | UnixFdsNotSupported
+
+-- | An authenticator that implements the D-Bus @EXTERNAL@ mechanism, which uses
+-- credential passing over a Unix socket, with support for Unix file descriptor passing.
+authenticatorWithUnixFds :: Authenticator SocketTransport
+authenticatorWithUnixFds = authExternal UnixFdsSupported
+
 -- | Implements the D-Bus @EXTERNAL@ mechanism, which uses credential
--- passing over a Unix socket.
-authExternal :: Authenticator SocketTransport
-authExternal = authenticator
-    { authenticatorClient = clientAuthExternal
-    , authenticatorServer = serverAuthExternal
+-- passing over a Unix socket, optionally supporting Unix file descriptor passing.
+authExternal :: UnixFdSupport -> Authenticator SocketTransport
+authExternal unixFdSupport = authenticator
+    { authenticatorClient = clientAuthExternal unixFdSupport
+    , authenticatorServer = serverAuthExternal unixFdSupport
     }
 
-clientAuthExternal :: SocketTransport -> IO Bool
-clientAuthExternal t = do
-    transportPut t (Data.ByteString.pack [0])
+clientAuthExternal :: UnixFdSupport -> SocketTransport -> IO Bool
+clientAuthExternal unixFdSupport t = do
+    transportPut t (Data.ByteString.pack [0]) []
     uid <- System.Posix.User.getRealUserID
     let token = concatMap (printf "%02X" . ord) (show uid)
     transportPutLine t ("AUTH EXTERNAL " ++ token)
     resp <- transportGetLine t
     case splitPrefix "OK " resp of
         Just _ -> do
-            transportPutLine t "BEGIN"
-            return True
+            ok <- do
+                case unixFdSupport of
+                    UnixFdsSupported -> do
+                        transportPutLine t "NEGOTIATE_UNIX_FD"
+                        respFd <- transportGetLine t
+                        return (respFd == "AGREE_UNIX_FD")
+                    UnixFdsNotSupported -> do
+                        return True
+            if ok then do
+                transportPutLine t "BEGIN"
+                return True
+            else
+                return False
         Nothing -> return False
 
-serverAuthExternal :: SocketTransport -> UUID -> IO Bool
-serverAuthExternal t uuid = do
-    let waitForBegin = do
-            resp <- transportGetLine t
-            if resp == "BEGIN"
-                then return ()
-                else waitForBegin
+serverAuthExternal :: UnixFdSupport -> SocketTransport -> UUID -> IO Bool
+serverAuthExternal unixFdSupport t uuid = do
+    let negotiateFdsAndBegin = do
+            line <- transportGetLine t
+            case line of
+                "NEGOTIATE_UNIX_FD" -> do
+                    let msg = case unixFdSupport of
+                            UnixFdsSupported ->
+                                "AGREE_UNIX_FD"
```

Comment:
This is pretty dumb: we send `AGREE_UNIX_FDS` even if the socket is not a unix socket, or if we are on windows where unix sockets don't support FD passing.

I think this is justified because the server is probably not that important, since mostly dbus clients will be connecting to a bus (which presumably will implement the checks properly), rather than implementing the server side of the connection. I'd be happy to change this, though.